pkgs/os-specific/linux/nixos-rebuild/nixos-rebuild.sh

```diff
@@ -503,11 +516,85 @@ if [ "$action" = dry-build ]; then
     extraBuildFlags+=(--dry-run)
 fi
 
+if [ "$action" = list-generations ]; then
+    if [ ! -L "$profile" ]; then
+        log "No profile \`$(basename "$profile")' found" >&2
+        exit 1
+    fi
+
+    generation_from_dir() {
+        generation_dir="$1"
+        generation_base="$(basename "$generation_dir")" # Has the format "system-123-link" for generation 123
+        no_link_gen="${generation_base%-link}"  # remove the "-link"
+        echo "${no_link_gen##*-}" # remove everything before the last dash
+    }
+    describe_generation(){
+        generation_dir="$1"
+        generation_number="$(generation_from_dir "$generation_dir")"
+        nixos_version="$(cat "$generation_dir/nixos-version" 2> /dev/null || echo "Unknown")"
+
+        kernel_dir="$(dirname "$(realpath "$generation_dir/kernel")")"
+        kernel_version="$(ls "$kernel_dir/lib/modules" || echo Unknown)"
+
+        configurationRevision="$("$generation_dir/sw/bin/nixos-version" --configurationRevision 2> /dev/null || :)"
+
+        # Old nixos-version output ignored unknown flags and just printed the version
+        # therefore the following workaround is done not to show the default output
+        nixos_version_default="$("$generation_dir/sw/bin/nixos-version")"
+        if [ "$configurationRevision" = "$nixos_version_default" ]; then
+             configurationRevision=""
+        fi
+
+        if [ -z "$json" ]; then
+            build_date="$(date --date="@$(stat "$generation_dir" --format=%W)" "+%a %F %T")"
+        else
+            # jq automatically quotes the output => don't try to quote it in output!
+            build_date="$(stat "$generation_dir" --format=%W | jq 'todate')"
+        fi
+
+        if [ -z "$json" ]; then
+            unset current_generation_tag
+        else
+            current_generation_tag="false"
+        fi
+        if [ "$(basename "$generation_dir")" = "$(readlink "$profile")" ]; then
+            if [ -z "$json" ]; then
+                current_generation_tag="  (current)"
+            else
+                current_generation_tag="true"
+            fi
+        fi
+
+        if [ -z $json ]; then
+            echo "$generation_number,$build_date,$nixos_version,$kernel_version,$configurationRevision$current_generation_tag"
+        else
+            # Escape userdefined strings
+            nixos_version="$(jq -aR <<< "$nixos_version")"
+            kernel_version="$(jq -aR <<< "$kernel_version")"
+            configurationRevision="$(jq -aR <<< "$configurationRevision")"
+            echo "{ \"generation\": $generation_number, \"date\": $build_date, \"nixosVersion\": $nixos_version, \"kernelVersion\": $kernel_version, \"configurationRevision\": $configurationRevision, \"current\": $current_generation_tag}"
```

Comment:
Not sure if we could add a linebreak here somewhere.

Reply:
Maybe add a second `echo`.

Reply:
Yes - this line is completely incomprehensible... Maybe one could even put each entry in a separate line or is that too much?

Reply:
I now put it in separate lines joined with backticks - what do you think about that?